peers_beacon/peers_beacon.cpp

```diff
@@ -0,0 +1,283 @@
+        }
+        return this->fetch_from_file();
+    }
+
+    throw std::runtime_error("no acceptable source for peers configured; must specify file or url or esr");
+}
+
+bool
+peers_beacon::fetch_from_file()
+{
+    auto filename = this->options->get_bootstrap_peers_file();
+    std::ifstream file(filename);
+    if (file.fail())
+    {
+        LOG(error) << "Failed to read bootstrap peers file " << filename;
+        return false;
+    }
+
+    LOG(info) << "Reading peers from " << filename;
+
+    return parse_and_save_peers(file);
+}
+
+bool
+peers_beacon::fetch_from_url()
+{
+    auto url = this->options->get_bootstrap_peers_url();
+    std::string peers = this->utils->sync_req(url);
+
+    LOG(info) << "Downloaded peer list from " << url;
+
+    std::stringstream stream;
+    stream << peers;
+
+    return parse_and_save_peers(stream);
+}
+
+bool
+peers_beacon::fetch_from_esr()
+{
+    auto swarm_id = this->options->get_swarm_id();
+    auto esr_address = this->options->get_swarm_info_esr_address();
+    auto esr_url = this->options->get_swarm_info_esr_url();
+
+    peers_list_t new_peers;
+    auto peer_ids = this->utils->get_peer_ids(swarm_id, esr_address, esr_url);
+    for (const auto& peer_id : peer_ids)
+    {
+        bzn::peer_address_t peer_info{this->utils->get_peer_info(swarm_id, peer_id, esr_address, esr_url)};
+        if (peer_info.host.empty()
+            || peer_info.port == 0
+            //|| peer_info.name.empty() // is it important that a peer have a name?
+            || peer_info.uuid.empty()
+                )
+        {
+            LOG(warning) << "Invalid peer information found in esr contract, ignoring info for peer: " << peer_id << " in swarm: " << swarm_id;
+        }
+        else
+        {
+            new_peers.emplace(peer_info);
+        }
+    }
+
+    return this->switch_peers_list(new_peers);
+}
+
+bool
+peers_beacon::switch_peers_list(const peers_list_t& new_peers)
+{
+    if (new_peers.size() == 0)
+    {
+        LOG(error) << "Failed to read any peers";
+        if (this->internal_current->size() > 0)
+        {
+            LOG(error) << "Keeping old peer list";
+        }
+        else
+        {
+            LOG(error) << "Old peers list also empty";
+        }
+        return false;
+    }
+
+    if(*(this->internal_current) != new_peers)
+    {
+        LOG(info) << "Switching to new peers list with " << new_peers.size() << " peers";
+    }
+
+    auto new_ordered = std::make_shared<ordered_peers_list_t>();
+    std::for_each(new_peers.begin(), new_peers.end(),
+        [&](const auto& peer)
+        {
+            new_ordered->push_back(peer);
+        });
+
+    std::sort(new_ordered->begin(), new_ordered->end(),
+        [](const auto& peer1, const auto& peer2)
+        {
+            return peer1.uuid.compare(peer2.uuid) < 0;
+        });
+
+    this->internal_current = std::make_shared<const peers_list_t>(new_peers);
+    this->internal_current_ordered = new_ordered;
+
```

Comment:
I could be wrong but it seems like this isn't thread-safe. The same shared_ptr instance (internal_current, internal_current_ordered) could be written here and read by current() / ordered() simultaneously while the shared_ptr object was in an inconsistent state.

Reply:
I had previously looked at the documentation and concluded that it was thread safe as is, but now that I look more closely I think you're actually right here and it technically should be protected. Shared pointers are apparently highly weird. I'll add a mutex here probably.